src/libpmemobj/critnib.c

```diff
@@ -0,0 +1,462 @@
+	struct critnib_node *n = c->deleted_node;
+	c->deleted_node = n->child[0];
+	return n;
+}
+
+/*
+ * crinib_insert -- write a key:value pair to the critnib structure
+ *
+ * Returns:
+ *  • 0 on success
+ *  • EINVAL if such a key already exists
+ *  • ENOMEM if we're out of memory
+ *
+ * Takes a global write lock but doesn't stall any readers.
+ */
+int
+critnib_insert(struct critnib *c, uint64_t key, void *value)
+{
+	os_mutex_lock(&c->mutex);
+	struct critnib_node *k = alloc_node(c);
+	if (!k)
+		return UNLOCK, ENOMEM;
+	k->path = key;
+	k->shift = ENDBIT;
+	k->child[0] = value;
+
+	struct critnib_node *n = c->root;
+	if (n == &nullnode) {
+		c->root = k;
+		return UNLOCK, 0;
+	}
+
+	struct critnib_node **parent = &c->root, *prev = c->root;
+
+	while (n->shift != ENDBIT && (key & (~NIB << n->shift)) == n->path) {
+		prev = n;
+		parent = &n->child[(key >> n->shift) & NIB];
+		n = *parent;
+	}
+
+	if (n == &nullnode) {
+		n = prev;
+		util_atomic_store_explicit64(
+			(uint64_t *)&n->child[(key >> n->shift) & NIB],
+			(uint64_t)k, memory_order_release);
+		return UNLOCK, 0;
+	}
+
+	uint64_t at = n->path ^ key;
+	if (!at) {
+		ASSERTeq(n->shift, ENDBIT);
+		free_node(c, k);
+		/* fail instead of replacing */
+		return UNLOCK, EINVAL;
+	}
+	int32_t sh = util_mssb_index64(at) & ~(SLICE - 1);
+
+	struct critnib_node *m = alloc_node(c);
+	if (!m) {
+		free_node(c, k);
+		return UNLOCK, ENOMEM;
+	}
+
+	for (int i = 0; i < SLNODES; i++)
+		m->child[i] = &nullnode;
+	uint64_t dir = (key >> sh) & NIB;
+	m->child[dir] = k;
+	m->child[(n->path >> sh) & NIB] = n;
+	m->shift = sh;
+	m->path = key & (~NIB << sh);
+	util_atomic_store_explicit64((uint64_t *)parent, (uint64_t)m,
+		memory_order_release);
+
+	return UNLOCK, 0;
+}
+
+/*
+ * critnib_remove -- delete a key from the critnib structure, return its value
+ */
+void *
+critnib_remove(struct critnib *c, uint64_t key)
+{
+	os_mutex_lock(&c->mutex);
+
+	struct critnib_node *n = c->root;
+	if (!n)
+		return UNLOCK, NULL;
+	uint64_t del = util_fetch_and_add64(&c->write_status, 1) % DELETED_LIFE;
+	free_node(c, c->pending_dels[del][0]);
+	free_node(c, c->pending_dels[del][1]);
+	c->pending_dels[del][1] = c->pending_dels[del][0] = 0;
+	if (n->shift == ENDBIT) {
+		if (n->path == key && n != &nullnode) {
+			util_atomic_store_explicit64((uint64_t *)&c->root,
+				(uint64_t)&nullnode, memory_order_release);
+			void *value = n->child[0];
+			c->pending_dels[del][0] = n;
+			return UNLOCK, value;
+		}
+		return UNLOCK, NULL;
+	}
+	struct critnib_node **k_parent = &c->root, **n_parent = &c->root,
+			    *k = n;
+
+	while (k->shift != ENDBIT) {
+		n_parent = k_parent;
+		n = k;
+		k_parent = &k->child[(key >> k->shift) & NIB];
+		k = *k_parent;
+	}
+	if (k->path != key)
+		return UNLOCK, NULL;
+
+	util_atomic_store_explicit64(
+		(uint64_t *)&n->child[(key >> n->shift) & NIB],
+		(uint64_t)&nullnode, memory_order_release);
+
+	int ochild = -1;
+	for (int i = 0; i < SLNODES; i++)
```

Comment:
I got lost here.
Is this here so that you remove parent nodes with only one remaining node?
Please add a comment, and also describe what exactly are the 'k' and 'n' variables.